bookserver/routers/rslogging.py

```diff
@@ -86,4 +107,75 @@ def set_tz_offset(
     values["tz_offset"] = tzreq.timezoneoffset
     response.set_cookie(key="RS_info", value=str(json.dumps(values)))
     rslogger.debug("setting timezone offset in session %s hours" % tzreq.timezoneoffset)
-    return "done"
+
+    return make_json_response()
+
+
+# runlog endpoint
+# ---------------
+# The `logRunEvent` client-side function calls this endpoint to record an activecode run
+@router.post("/runlog")
+async def runlog(request: Request, response: Response, data: LogRunIncoming):
+    # First add a useinfo entry for this run
+    if request.state.user:
+        if data.course != request.state.user.course_name:
+            return make_json_response(
+                status=status.HTTP_401_UNAUTHORIZED,
+                detail="You appear to have changed courses in another tab.  Please switch to this course",
+            )
+        data.sid = request.state.user.username
+    else:
+        if data.clientLoginStatus == "true":
+            rslogger.error("Session Expired")
+            return make_json_response(
+                status=status.HTTP_401_UNAUTHORIZED, detail="Session Expired"
+            )
+        else:
+            return make_json_response(status=status.HTTP_401_UNAUTHORIZED)
+
+    # everything after this assumes that the user is logged in
+
+    useinfo_dict = data.dict()
+    useinfo_dict["course_id"] = useinfo_dict.pop("course")
+    useinfo_dict["timestamp"] = datetime.utcnow()
+    if data.errinfo != "success":
+        useinfo_dict["event"] = "ac_error"
+        useinfo_dict["act"] = str(data.errinfo)[:512]
+    else:
+        useinfo_dict["act"] = "run"
+        if "event" not in useinfo_dict:
+            useinfo_dict["event"] = "activecode"
+
+    await create_useinfo_entry(UseinfoValidation(**useinfo_dict))
```

Comment:
Subtle, but nice: this does secondary validation as a part of creating the useinfo entry.